File: vscodeExtensions/liberrpa-project-manager/src/componentManagement/protocol.ts
// FileName: protocol.ts

import type { DictProjectManifestV1 } from "../projectManifest";

// align protocol types with Python.
export type Str_ProjectType = "flow" | "component";

export interface DictComponentManagementWarning {
  code: string;
  message: string;
  details?: Record<string, unknown>;

  file?: string;
  line?: number;
  functionName?: string;
  snippetKey?: string;

  [key: string]: unknown;
}

export interface DictProtocolDependencyOperation_Add {
  operation: "addComponentDependency";
  componentId: string;
  requirement: string;
}

export interface DictProtocolDependencyOperation_Update {
  operation: "updateComponents";
  componentIds: string[];
}

export interface DictProtocolDependencyOperation_ChangeRequirement {
  operation: "changeComponentRequirement";
  componentId: string;
  requirement: string;
}

export interface DictProtocolDependencyOperation_Remove {
  operation: "removeComponentDependency";
  componentId: string;
}

export type DictProtocolDependencyOperation =
  | DictProtocolDependencyOperation_Add
  | DictProtocolDependencyOperation_Update
  | DictProtocolDependencyOperation_ChangeRequirement
  | DictProtocolDependencyOperation_Remove;

export interface DictProtocolRequest_PublishComponent {
  schemaVersion: 1;
  operation: "publishComponent";
  projectPath: string;
}

export interface DictProtocolRequest_RebuildRepositoryIndex {
  schemaVersion: 1;
  operation: "rebuildRepositoryIndex";
}

export interface DictProtocolRequest_ImportComponentWheels {
  schemaVersion: 1;
  operation: "importComponentWheels";
  wheelPaths: string[];
}

export interface DictProtocolRequest_GetComponentRepositoryCatalog {
  schemaVersion: 1;
  operation: "getComponentRepositoryCatalog";
}

export interface DictProtocolRequest_GetProjectDependencyState {
  schemaVersion: 1;
  operation: "getProjectDependencyState";
  projectPath: string;
}

export interface DictProtocolRequest_BuildProjectDependencyPlan {
  schemaVersion: 1;
  operation: "buildProjectDependencyPlan";
  projectPath: string;
  dependencyOperation: DictProtocolDependencyOperation;
}

export interface DictProtocolRequest_ApplyProjectDependencyPlan {
  schemaVersion: 1;
  operation: "applyProjectDependencyPlan";
  projectPath: string;
  dependencyOperation: DictProtocolDependencyOperation;
  confirmedPlanSha256: string;
}

export interface DictProtocolRequest_RepairProjectComponents {
  schemaVersion: 1;
  operation: "repairProjectComponents";
  projectPath: string;
}

export type DictProtocolRequest =
  | DictProtocolRequest_PublishComponent
  | DictProtocolRequest_RebuildRepositoryIndex
  | DictProtocolRequest_ImportComponentWheels
  | DictProtocolRequest_GetComponentRepositoryCatalog
  | DictProtocolRequest_GetProjectDependencyState
  | DictProtocolRequest_BuildProjectDependencyPlan
  | DictProtocolRequest_ApplyProjectDependencyPlan
  | DictProtocolRequest_RepairProjectComponents;

export interface DictProtocolResult_PublishBase {
  componentId: string;
  packageName: string;
  astSnippetsFile: string;
  snippetsJsoncFile: string;
  generatedCount: number;
  skippedCount: number;
  warningCount: number;
}

export interface DictProtocolResult_Publish_PreparationCreated extends DictProtocolResult_PublishBase {
  status: "preparationCreated";
}

export interface DictProtocolResult_Publish_Published extends DictProtocolResult_PublishBase {
  status: "published" | "alreadyPublished";
  version: string;
  wheelFile: string;
  sha256: string;
  excludedCount: number;
  handWrittenCount: number;
  finalCount: number;
}

export type DictProtocolResult_Publish =
  | DictProtocolResult_Publish_PreparationCreated
  | DictProtocolResult_Publish_Published;

export interface DictProtocolResult_RepositoryIndexRebuilt {
  status: "repositoryIndexRebuilt";
  componentCount: number;
  versionCount: number;
}

export interface DictProtocolResult_ComponentWheelsImported_Component {
  sourcePath: string;
  componentId: string;
  packageName: string;
  version: string;
  wheelFile: string;
  sha256: string;
  status: "imported" | "alreadyImported";
}

export interface DictProtocolResult_ComponentWheelsImported {
  status: "componentWheelsImported";
  importedCount: number;
  alreadyImportedCount: number;
  components: DictProtocolResult_ComponentWheelsImported_Component[];
}

export interface DictRepository_ComponentVersion {
  version: string;
  displayName: string;
  description: string;
  manifestSchemaVersion: 1;
  wheelFile: string;
  sha256: string;
  requiresLiberrpa: string;
  componentDependencies: Record<string, string>;
}

export interface DictProtocolResult_RepositoryCatalog_Component {
  componentId: string;
  packageName: string;
  versions: DictRepository_ComponentVersion[];
}

export interface DictProtocolResult_RepositoryCatalog {
  status: "componentRepositoryCatalog";
  repositoryPath: string;
  componentCount: number;
  versionCount: number;
  components: DictProtocolResult_RepositoryCatalog_Component[];
}

export interface DictComponentsLock_Root_FlowProject {
  manifestFile: "flow.json";
  manifestSchemaVersion: 1;
  requiresLiberrpa: string;
  componentDependencies: Record<string, string>;
  resolutionInputSha256: string;
}

export interface DictComponentsLock_Root_ComponentProject {
  manifestFile: "component.json";
  manifestSchemaVersion: 1;
  componentId: string;
  packageName: string;
  requiresLiberrpa: string;
  componentDependencies: Record<string, string>;
  resolutionInputSha256: string;
}

export type DictComponentsLock_Root =
  | DictComponentsLock_Root_FlowProject
  | DictComponentsLock_Root_ComponentProject;

export interface DictComponentsLock_Component {
  manifestSchemaVersion: 1;
  packageName: string;
  displayName: string;
  version: string;
  wheelFile: string;
  sha256: string;
  requiresLiberrpa: string;
  componentDependencies: Record<string, string>;
}

export interface DictComponentsLock_File {
  schemaVersion: 1;
  root: DictComponentsLock_Root;
  components: Record<string, DictComponentsLock_Component>;
}

export type Str_ProjectDependency_LockState =
  | "notRequired"
  | "missing"
  | "invalid"
  | "stale"
  | "valid";

export type Str_ProjectDependency_ComponentsState =
  | "notRequired"
  | "unverified"
  | "missing"
  | "damaged"
  | "valid";

export type Str_ProjectDependency_EnvironmentState =
  | "unknown"
  | "compatible"
  | "incompatible";

export type Str_ProjectDependency_RepairState =
  | "notApplicable"
  | "available"
  | "repositoryUnavailable"
  | "wheelMissing"
  | "wheelHashMismatch";

export interface DictProjectDependency_DirectChange {
  componentId: string;
  change: "added" | "removed" | "requirementChanged";
  previousRequirement?: string;
  targetRequirement?: string;
}

export interface DictProjectDependency_ResolvedChange {
  componentId: string;
  packageName: string;
  displayName: string;
  change: "added" | "removed" | "upgraded" | "downgraded";
  previousVersion?: string;
  targetVersion?: string;
}

export interface DictProtocolResult_ComponentsFolder {
  componentsPath: string;
  componentCount: number;
  fileCount: number;
}

export interface DictProtocolResult_ProjectDependencyState {
  status: "projectDependencyState";
  projectPath: string;
  projectType: Str_ProjectType;
  manifest: DictProjectManifestV1;
  componentsLock: DictComponentsLock_File | null;
  lockState: Str_ProjectDependency_LockState;
  componentsState: Str_ProjectDependency_ComponentsState;
  environmentState: Str_ProjectDependency_EnvironmentState;
  repairState: Str_ProjectDependency_RepairState;
  details: Record<string, unknown>;
}

export interface DictProtocolResult_ProjectDependencyPlan {
  status: "projectDependencyPlanCreated";
  planSha256: string;
  sourceManifest: DictProjectManifestV1;
  sourceComponentsLock: DictComponentsLock_File | null;
  targetManifest: DictProjectManifestV1;
  targetComponentsLock: DictComponentsLock_File | null;
  directDependencyChanges: DictProjectDependency_DirectChange[];
  resolvedComponentChanges: DictProjectDependency_ResolvedChange[];
}

export interface DictProtocolResult_ProjectDependencyPlanApplied {
  status: "projectDependencyPlanApplied";
  planSha256: string;
  targetManifest: DictProjectManifestV1;
  targetComponentsLock: DictComponentsLock_File | null;
  directDependencyChanges: DictProjectDependency_DirectChange[];
  resolvedComponentChanges: DictProjectDependency_ResolvedChange[];
  componentsFolder: DictProtocolResult_ComponentsFolder | null;
}

export interface DictProtocolResult_ProjectComponentsRepaired {
  status: "projectComponentsRepaired";
  componentsFolder: DictProtocolResult_ComponentsFolder;
}

export interface DictProtocolError {
  code: string;
  message: string;
  details: Record<string, unknown>;
}

export interface DictProtocolSuccess_Raw {
  schemaVersion: 1;
  ok: true;
  result: Record<string, unknown>;
  warnings: DictComponentManagementWarning[];
}

export interface DictProtocolResponse_Error {
  schemaVersion: 1;
  ok: false;
  error: DictProtocolError;
}

export type DictProtocolResponse_Raw = DictProtocolSuccess_Raw | DictProtocolResponse_Error;

File: condaLibrary/liberrpa/ComponentManagement/Types/_Protocol.py
# FileName: _Protocol.py
__author__ = "Jiyan Hu"
__email__ = "mailwork.hu@gmail.com"
__license__ = "GNU Affero General Public License v3.0 or later"
__copyright__ = f"Copyright (C) 2025 {__author__}"


from liberrpa.ComponentManagement.Types._Warning import DictComponentManagementWarning
from liberrpa.ComponentManagement.Types._Manifest import Str_ProjectType
from liberrpa.ComponentManagement.Types._Components import DictComponentsLock_File
from liberrpa.ComponentManagement.Types._Repository import (
    DictRepository_ComponentVersionEntry,
)
from liberrpa.ComponentManagement.Types._Dependency import (
    DictProjectDependency_DirectChange,
    DictProjectDependency_ResolvedChange,
    Str_ProjectDependency_LockState,
    Str_ProjectDependency_ComponentsState,
    Str_ProjectDependency_EnvironmentState,
    Str_ProjectDependency_RepairState,
)


from typing import Literal, TypedDict


class DictProtocolDependencyOperation_Add(TypedDict):
    operation: Literal["addComponentDependency"]
    componentId: str
    requirement: str


class DictProtocolDependencyOperation_Update(TypedDict):
    operation: Literal["updateComponents"]
    componentIds: list[str]


class DictProtocolDependencyOperation_Resolve(TypedDict):
    operation: Literal["resolveProjectDependencies"]


class DictProtocolDependencyOperation_ChangeRequirement(TypedDict):
    operation: Literal["changeComponentRequirement"]
    componentId: str
    requirement: str


class DictProtocolDependencyOperation_Remove(TypedDict):
    operation: Literal["removeComponentDependency"]
    componentId: str


type DictProtocolDependencyOperation = (
    DictProtocolDependencyOperation_Add
    | DictProtocolDependencyOperation_Update
    | DictProtocolDependencyOperation_Resolve
    | DictProtocolDependencyOperation_ChangeRequirement
    | DictProtocolDependencyOperation_Remove
)


class DictProtocolRequest_PublishComponent(TypedDict):
    schemaVersion: Literal[1]
    operation: Literal["publishComponent"]

    projectPath: str


class DictProtocolRequest_ImportComponentWheels(TypedDict):
    schemaVersion: Literal[1]
    operation: Literal["importComponentWheels"]

    wheelFilePaths: list[str]


class DictProtocolRequest_RebuildRepositoryIndex(TypedDict):
    schemaVersion: Literal[1]
    operation: Literal["rebuildRepositoryIndex"]


class DictProtocolRequest_GetComponentRepositoryCatalog(TypedDict):
    schemaVersion: Literal[1]
    operation: Literal["getComponentRepositoryCatalog"]


class DictProtocolRequest_GetProjectManifestDefaults(TypedDict):
    schemaVersion: Literal[1]
    operation: Literal["getProjectManifestDefaults"]


class DictProtocolRequest_GetProjectDependencyState(TypedDict):
    schemaVersion: Literal[1]
    operation: Literal["getProjectDependencyState"]

    projectPath: str


class DictProtocolRequest_ValidatePackagedFlowProject(TypedDict):
    schemaVersion: Literal[1]
    operation: Literal["validatePackagedFlowProject"]

    projectPath: str


class DictProtocolRequest_BuildProjectDependencyPlan(TypedDict):
    schemaVersion: Literal[1]
    operation: Literal["buildProjectDependencyPlan"]

    projectPath: str
    dependencyOperation: DictProtocolDependencyOperation


class DictProtocolRequest_ApplyProjectDependencyPlan(TypedDict):
    schemaVersion: Literal[1]
    operation: Literal["applyProjectDependencyPlan"]

    projectPath: str
    dependencyOperation: DictProtocolDependencyOperation
    confirmedPlanSha256: str


class DictProtocolRequest_RepairProjectComponents(TypedDict):
    schemaVersion: Literal[1]
    operation: Literal["repairProjectComponents"]

    projectPath: str


type DictProtocolRequest = (
    DictProtocolRequest_PublishComponent
    | DictProtocolRequest_ImportComponentWheels
    | DictProtocolRequest_RebuildRepositoryIndex
    | DictProtocolRequest_GetComponentRepositoryCatalog
    | DictProtocolRequest_GetProjectManifestDefaults
    | DictProtocolRequest_GetProjectDependencyState
    | DictProtocolRequest_ValidatePackagedFlowProject
    | DictProtocolRequest_BuildProjectDependencyPlan
    | DictProtocolRequest_ApplyProjectDependencyPlan
    | DictProtocolRequest_RepairProjectComponents
)


class _DictProtocolResult_PublishBase(TypedDict):
    componentId: str
    packageName: str

    astSnippetsFile: str
    snippetsJsoncFile: str
    generatedCount: int
    skippedCount: int
    warningCount: int


class _DictProtocolResult_Publish_PreparationCreated(_DictProtocolResult_PublishBase):
    status: Literal["preparationCreated"]


class _DictProtocolResult_Publish_Published(_DictProtocolResult_PublishBase):
    status: Literal["published", "alreadyPublished"]

    version: str

    excludedCount: int
    handWrittenCount: int
    finalCount: int

    wheelFileName: str
    sha256: str


type DictProtocolResult_Publish = (
    _DictProtocolResult_Publish_PreparationCreated | _DictProtocolResult_Publish_Published
)


class DictProtocolResult_ComponentWheelsImported_Component(TypedDict):
    # It has the same values with Info_Repository_Import_ComponentResult.
    sourceWheelFilePath: str

    componentId: str
    packageName: str
    version: str

    wheelFileName: str
    sha256: str

    status: Literal["imported", "alreadyImported"]


class DictProtocolResult_ComponentWheelsImported(TypedDict):
    status: Literal["componentWheelsImported"]
    importedCount: int
    alreadyImportedCount: int
    components: list[DictProtocolResult_ComponentWheelsImported_Component]


class DictProtocolResult_RepositoryIndexRebuilt(TypedDict):
    status: Literal["repositoryIndexRebuilt"]
    componentCount: int
    versionCount: int


class DictProtocolResult_RepositoryCatalog_Component(TypedDict):
    componentId: str
    packageName: str
    versions: list[DictRepository_ComponentVersionEntry]


class DictProtocolResult_RepositoryCatalog(TypedDict):
    status: Literal["componentRepositoryCatalog"]
    repositoryPath: str
    componentCount: int
    versionCount: int
    components: list[DictProtocolResult_RepositoryCatalog_Component]


class DictProtocolResult_ProjectManifestDefaults(TypedDict):
    status: Literal["projectManifestDefaults"]
    installedLiberrpaVersion: str
    requiresLiberrpa: str


class DictProtocolResult_ComponentsFolder(TypedDict):
    componentsFolderPath: str
    componentCount: int
    fileCount: int


class DictProtocolResult_ProjectDependencyState(TypedDict):
    status: Literal["projectDependencyState"]
    projectPath: str
    projectType: Str_ProjectType
    manifest: dict[str, object]
    componentsLock: DictComponentsLock_File | None
    lockState: Str_ProjectDependency_LockState
    componentsState: Str_ProjectDependency_ComponentsState
    environmentState: Str_ProjectDependency_EnvironmentState
    repairState: Str_ProjectDependency_RepairState
    details: dict[str, object]


class DictProtocolResult_PackagedFlowProjectValidated(TypedDict):
    status: Literal["packagedFlowProjectValidated"]


class DictProtocolResult_ProjectDependencyPlan(TypedDict):
    status: Literal["projectDependencyPlanCreated"]
    planSha256: str
    sourceManifest: dict[str, object]
    sourceComponentsLock: DictComponentsLock_File | None
    targetManifest: dict[str, object]
    targetComponentsLock: DictComponentsLock_File | None
    directDependencyChanges: list[DictProjectDependency_DirectChange]
    resolvedComponentChanges: list[DictProjectDependency_ResolvedChange]


class DictProtocolResult_ProjectDependencyPlanApplied(TypedDict):
    status: Literal["projectDependencyPlanApplied"]
    planSha256: str
    targetManifest: dict[str, object]
    targetComponentsLock: DictComponentsLock_File | None
    directDependencyChanges: list[DictProjectDependency_DirectChange]
    resolvedComponentChanges: list[DictProjectDependency_ResolvedChange]
    componentsFolder: DictProtocolResult_ComponentsFolder | None


class DictProtocolResult_ProjectComponentsRepaired(TypedDict):
    status: Literal["projectComponentsRepaired"]
    componentsFolder: DictProtocolResult_ComponentsFolder


type DictProtocolResult = (
    DictProtocolResult_Publish
    | DictProtocolResult_ComponentWheelsImported
    | DictProtocolResult_RepositoryIndexRebuilt
    | DictProtocolResult_RepositoryCatalog
    | DictProtocolResult_ProjectManifestDefaults
    | DictProtocolResult_ProjectDependencyState
    | DictProtocolResult_PackagedFlowProjectValidated
    | DictProtocolResult_ProjectDependencyPlan
    | DictProtocolResult_ProjectDependencyPlanApplied
    | DictProtocolResult_ProjectComponentsRepaired
)


class DictProtocolError(TypedDict):
    code: str
    message: str
    details: dict[str, object]


class DictProtocolSuccess_PublishComponent(TypedDict):
    schemaVersion: Literal[1]
    ok: Literal[True]
    result: DictProtocolResult_Publish
    warnings: list[DictComponentManagementWarning]


class DictProtocolSuccess_ComponentWheelsImported(TypedDict):
    schemaVersion: Literal[1]
    ok: Literal[True]
    result: DictProtocolResult_ComponentWheelsImported
    warnings: list[DictComponentManagementWarning]


class DictProtocolSuccess_RepositoryIndexRebuilt(TypedDict):
    schemaVersion: Literal[1]
    ok: Literal[True]
    result: DictProtocolResult_RepositoryIndexRebuilt
    warnings: list[DictComponentManagementWarning]


class DictProtocolSuccess_RepositoryCatalog(TypedDict):
    schemaVersion: Literal[1]
    ok: Literal[True]
    result: DictProtocolResult_RepositoryCatalog
    warnings: list[DictComponentManagementWarning]


class DictProtocolSuccess_ProjectManifestDefaults(TypedDict):
    schemaVersion: Literal[1]
    ok: Literal[True]
    result: DictProtocolResult_ProjectManifestDefaults
    warnings: list[DictComponentManagementWarning]


class DictProtocolSuccess_ProjectDependencyState(TypedDict):
    schemaVersion: Literal[1]
    ok: Literal[True]
    result: DictProtocolResult_ProjectDependencyState
    warnings: list[DictComponentManagementWarning]


class DictProtocolSuccess_PackagedFlowProjectValidated(TypedDict):
    schemaVersion: Literal[1]
    ok: Literal[True]
    result: DictProtocolResult_PackagedFlowProjectValidated
    warnings: list[DictComponentManagementWarning]


class DictProtocolSuccess_ProjectDependencyPlan(TypedDict):
    schemaVersion: Literal[1]
    ok: Literal[True]
    result: DictProtocolResult_ProjectDependencyPlan
    warnings: list[DictComponentManagementWarning]


class DictProtocolSuccess_ProjectDependencyPlanApplied(TypedDict):
    schemaVersion: Literal[1]
    ok: Literal[True]
    result: DictProtocolResult_ProjectDependencyPlanApplied
    warnings: list[DictComponentManagementWarning]


class DictProtocolSuccess_ProjectComponentsRepaired(TypedDict):
    schemaVersion: Literal[1]
    ok: Literal[True]
    result: DictProtocolResult_ProjectComponentsRepaired
    warnings: list[DictComponentManagementWarning]


type DictProtocolSuccess = (
    DictProtocolSuccess_PublishComponent
    | DictProtocolSuccess_ComponentWheelsImported
    | DictProtocolSuccess_RepositoryIndexRebuilt
    | DictProtocolSuccess_RepositoryCatalog
    | DictProtocolSuccess_ProjectManifestDefaults
    | DictProtocolSuccess_ProjectDependencyState
    | DictProtocolSuccess_PackagedFlowProjectValidated
    | DictProtocolSuccess_ProjectDependencyPlan
    | DictProtocolSuccess_ProjectDependencyPlanApplied
    | DictProtocolSuccess_ProjectComponentsRepaired
)


class DictProtocolResponse_Error(TypedDict):
    schemaVersion: Literal[1]
    ok: Literal[False]
    error: DictProtocolError


type DictProtocolResponse = DictProtocolSuccess | DictProtocolResponse_Error
